libs/wingsdk/src/target-sim/function.inflight.ts

```diff
@@ -78,12 +96,94 @@ export class Function implements IFunctionClient, ISimulatorResourceInstance {
               sourceType: FUNCTION_TYPE,
               timestamp: new Date().toISOString(),
             });
-          });
+          };
         }
 
-        const index = vm.runFile(this.filename);
-        return index.handler(payload);
+        return runSandbox(this.bundle, payload, {
+          timeout: this.timeout,
+          context: {
+            process: sandboxProcess,
+            $simulator: this.context,
+            console: sandboxConsole,
+          },
+        });
       },
     });
   }
 }
+
+interface RunSandboxOptions {
+  readonly context: { [key: string]: any };
+  readonly timeout: number;
+}
+
+/**
+ * Runs user code in a sandboxed environment. The code is expected to export a `handler` async
+ * function which take a payload and returns a result via a promise.
+ *
+ * @param filepath A path to a bundled JavaScript file (no "require")
+ * @param payload The payload JSON object to pass to the handler
+ * @param opts Sandbox options
+ * @returns A promise
+ */
+async function runSandbox(
+  filepath: string,
+  payload: any,
+  opts: RunSandboxOptions
+): Promise<any> {
+  const ctx: any = {};
+
+  // create a copy of all the globals from our current context.
+  for (const k of Object.getOwnPropertyNames(global)) {
+    try {
+      ctx[k] = (global as any)[k];
+    } catch {
+      // ignore unresolvable globals (see https://github.com/winglang/wing/pull/1923)
+    }
+  }
+
+  // append the user's context
+  for (const k of Object.keys(opts.context)) {
+    ctx[k] = opts.context[k];
+  }
+
+  const code = await readFile(filepath, "utf-8");
+
+  return new Promise(($resolve, $reject) => {
+    const wrapper = [
+      "const exports = {};",
+      code,
+      `exports.handler(${JSON.stringify(
+        payload
+      )}).then($resolve).catch($reject);`,
+    ].join("\n");
+
+    const context = vm.createContext({
+      ...ctx,
+      $resolve,
+      $reject,
+      require, // to support requiring node.js sdk modules (others will be bundled)
+    });
+
+    // emit an explicit error when trying to access `__dirname` and `__filename` because we cannot
+    // resolve these when bundling (this is true both for simulator and the cloud since we are
+    // bundling there as well).
+    const forbidGlobal = (name: string) => {
+      Object.defineProperty(context, name, {
+        get: () => {
+          throw new Error(
+            `${name} cannot be used within bundled cloud functions`
+          );
+        },
+      });
+    };
+
+    forbidGlobal("__dirname");
+    forbidGlobal("__filename");
```

Comment:
Could we have `__dirname` and `__filename` return the directory/filename of the final bundle path produced by esbuild (assuming we're bundling everything into a single file)?

Reply:
Perhaps it makes sense to set `__filename` to point to the bundle file, but I think it will likely be a bad idea to make `__dirname` point to the bundle file directory because in most cases `__dirname` is used to locate a sibling file in a module's node.js package so it's almost comes with some assumptions about the structure of the package itself.

For now, I rather just emit a clear error and we will see if this needs tweaking in the future if you guys are okay with that.

Reply:
Totally cool with me 🙂